{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<center>\n",
    "    \n",
    "<h1>An example of the Gilepsie Algorithm: the Hok/Sok mechanism.</h1>\n",
    "</center>\n",
    "\n",
    "We propose in this project to model a simplify the Hok/sok system. We will propose to code a stochastic simulator that will represent the production of both Hok and Sok mRNAs and Hok proteins.\n",
    "\n",
    "## The Hok/sok system\n",
    "\n",
    "Text adapted from from wikipedia (https://en.wikipedia.org/wiki/Hok/sok_system):\n",
    "\n",
    "<i>\n",
    "The hok/sok system is a postsegregational killing mechanism employed by the R1 plasmid in Escherichia coli. [â€¦] The hok/sok system involves two genes:\n",
    "- hok, host killing : RNA with a long lifetime, its protein is toxic\n",
    "- sok, suppression of killing : RNA with a short lifetime, RNA antitoxin\n",
    "\n",
    "When E. coli undergoes cell division, the two daughter cells inherit the long-lived hok toxin from the parent cell. Due to the short half-life of the sok antitoxin, daughter cells inherit only small amounts and it quickly degrades.\n",
    "\n",
    "If a daughter cell has inherited the R1 plasmid, it has inherited the sok gene and a strong promoter which brings about high levels of transcription. So much so that in an R1-positive cell, Sok transcript exists in considerable molar excess over Hok mRNA. Sok RNA then indirectly inhibits the translation of hok by inhibiting mok translation. There is a complementary region where sok transcript binds hok mRNA directly (pictured). Instead, sok RNA regulates the translation of the mok open reading frame, which nearly entirely overlaps that of hok. It is this translation-coupling which effectively allows sok RNA to repress the translation of hok mRNA.\n",
    "\n",
    "The sok transcript forms a duplex with the leader region of hok mRNA and this is recognized by RNase III and degraded. The cleavage products are very unstable and soon decay.\n",
    "\n",
    "With plasmid || When the plasmid leaves\n",
    "-||-\n",
    "<img src=\"./figures/wikipedia.svg\" alt=\"Transcription\" style=\"width: 400px;\"/> &nbsp;&nbsp;&nbsp;||&nbsp;&nbsp;&nbsp;  <img src=\"./figures/wikipedia1.svg\" alt=\"TwoStageModel1.png\" style=\"width: 400px;\"/>\n",
    "\n",
    "(image inspired by [this wikipedia diagram](https://en.wikipedia.org/wiki/Hok/sok_system#/media/File:Hok_sok_system_R1_plasmid_present.gif) )\n",
    "\n",
    "</i>\n",
    "\n",
    "***Some useful data***:\n",
    "\n",
    "Given parameters:\n",
    "- the reaction rate of the mRNA hok and sok is  $3 \\cdot 10^2 M^{-1} s^{-1}$ (Franch et al., 1997)\n",
    "- the typical E. Coli volume $1\\,\\mu m^3$\n",
    "\n",
    "Parameter you will have to search in the litterature:\n",
    "- hok mRNA half-life\n",
    "- sok RNA half-life\n",
    "\n",
    "Other information that are not easy to find in the literature. To begin with, we can propose that:\n",
    "- rate of hok mRNA creation $0.3\\,min^{-1}$\n",
    "- rate of sok RNA creation $3\\,min^{-1}$\n",
    "- rate of Hok protein creation per mRNA $0.3\\,min^{-1}$\n",
    "- rate of Hok protein proteolysis $0.01\\,min^{-1}$\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Let's play with random numbers\n",
    "\n",
    "As we will do stochastic simulations, we will need to play with random numbers. For that, we will import the scientific library **numpy** and the plotling library ***pyplot*** (we will use ***plt*** as a shortcut for ***pyplot***)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy\n",
    "import matplotlib.pyplot as plt\n",
    "# The next statement ensure that the figure are plot in the notebook\n",
    "%matplotlib inline "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There is a submodule of numpy that deals with random numbers: ***numpy.random***. From this, you have access to a series usual random distribution: \n",
    "\n",
    "- `numpy.random.random()` : returns a random float in $[0,1[$\n",
    "- `numpy.random.randint(N)` : returns a random integer between 0 and N (without including N)\n",
    "- `numpy.random.exponential(scale)` : returns a float that follow an exponentially distributed random distribution of parameter $1/scale$. ***Caution:*** for some reason, it is $1/scale$ and not $scale$. \n",
    "- etc.\n",
    "\n",
    "Let's try:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A random integer:  4\n",
      "A random integer:  0\n",
      "A random integer:  3\n",
      "A random integer:  0\n",
      "A random integer:  2\n",
      "A random integer:  0\n",
      "A random integer:  5\n",
      "A random integer:  9\n",
      "A random integer:  3\n",
      "A random integer:  9\n"
     ]
    }
   ],
   "source": [
    "for i in range(10):\n",
    "    print(\"A random integer: \",numpy.random.randint(10)) # prints a random integer between 0 and 10 (without including 10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "For each of these function, you can specify a size, in which case the function will return an ***array*** of the specify size. For instance, `numpy.random.randint(10,size=100)` will return a array of 100 integers randomly chosen between 0 and 10 (without including 10).\n",
    "\n",
    "***Exercise*** : Create an array of 1000 elements that follow a exponential random distribution of parameter 0.1 and plot its histogram.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### TO COMPLETE\n",
    "distribParam = 0.1\n",
    "randomArray = ???????????????????????\n",
    "\n",
    "####\n",
    "\n",
    "plt.hist(randomArray,normed=True,bins=30,alpha=.5,label='Random') # This statement plots the histogram\n",
    "X = numpy.linspace(0,70,100)\n",
    "plt.plot(X,distribParam * numpy.exp(-distribParam*X),label='Distribution') # This statement plots the distribution\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One important random distribution is missing in numpy.random . It is the *weighted choice*.\n",
    "\n",
    "\n",
    "Example: Let's imagine you have a the biased dice, such as you have:\n",
    "- 1/2 chance to obtain a 1\n",
    "- 1/10 chances to obtain a 2\n",
    "- 1/10 chances to obtain a 3\n",
    "- 1/10 chances to obtain a 4\n",
    "- 1/10 chances to obtain a 5\n",
    "- 1/10 chances to obtain a 6\n",
    "<img src=\"https://upload.wikimedia.org/wikipedia/commons/a/a5/6sided_dice.jpg\" alt=\"Dice\" style=\"width: 200px;\"/>\n",
    "<div style=\"text-align: right\">[image from wikipedia](https://upload.wikimedia.org/wikipedia/commons/a/a5/6sided_dice.jpg)</div>\n",
    "\n",
    "(where if you had a fair dice, the chances will be 1/6 for each of them).\n",
    "\n",
    "How to draw a sample from such random variable ? The function to draw it is not given in **numpy.random**. We will need it in the project, so it will be necessary to create this function ourself. Even if it is not complicated to code, I give it directly in the next cell.\n",
    "\n",
    "**Note**: this function do not specifically expects weights as a list of probabilities that sums to 1. If the sum of all the weights is not 1, it automatically normalized them. The function also returns the sum of all the weights (it will be useful later). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def randomWeightedChoice(weights,size=1):\n",
    "    \"\"\"\n",
    "    Randomly choice an index accorging to the given weights.\n",
    "    - weights: numpy 1D array like\n",
    "            The probability weights for each choice (if the sum of all weights \n",
    "            is not 1, we scale it to 1)\n",
    "    - size: int\n",
    "            The size of the sample\n",
    "    \n",
    "    Returns:\n",
    "    - index: int or 1DArray\n",
    "        The indexes of the choice (if size==1, returns an interger)\n",
    "    - summ: the sum of all the weights (before normalizing)\n",
    "    \"\"\"\n",
    "    bins = numpy.add.accumulate(weights,dtype=float)\n",
    "    summ = bins[-1]\n",
    "    \n",
    "    if size==1:\n",
    "        return  int(numpy.searchsorted(bins,summ*numpy.random.random_sample())),summ\n",
    "    else:\n",
    "        result = numpy.zeros(size,dtype=int)\n",
    "        samples = summ*numpy.random.random_sample(size=size)\n",
    "\n",
    "        for i,s in enumerate(samples):\n",
    "            result[i] = int(numpy.searchsorted(bins,s))\n",
    "            \n",
    "        return result,summ"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This function will be VERY USEFULL in the remaining of the project.\n",
    "\n",
    "**Exercice:** Let's check that this function works as expected. We will try the `randomWeightedChoice` on the previous example of the bias dice. Sample a large array with  `randomWeightedChoice` with the weights corresponding to the bias dice example. Check that the obtained histogram corresponds to the expected distribution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diceWeights = numpy.array([0.5,0.1,0.1,0.1,0.1,0.1])\n",
    "\n",
    "\n",
    "#### TO COMPLETE\n",
    "randomArray = ???????????????????????\n",
    "\n",
    "plt.hist(????????????????????,normed=True,bins=numpy.arange(7),\n",
    "                     alpha=.5,label='Random Choice') # This statement plots the histogram\n",
    "plt.plot(?????????????????????,label='Distribution') # This statement plots the distribution\n",
    "plt.legend()\n",
    "\n",
    "####\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gillepsie algorithm's main tool: exponential distributions\n",
    "\n",
    "This section contains the only maths that will be usefull to perform our simulations.\n",
    "\n",
    "The exponential distribution is a continuous distribution (i.e. it represents floats) and it often used to represent periods of time. Its probability density function is given by $$f(x)=a e^{-a\\cdot x}.$$\n",
    "\n",
    "Typically the parameter $a$ of the distribution given the **rate** at which happen the event. \n",
    " ***\n",
    "**Example:** For instance, let's model the mRNA degradation with a exponentially random variable, this mRNA degrates on average in 10 minutes, the parameter of the exponential dsitribution will then be $0.1\\,min^{-1}$.\n",
    "***\n",
    "\n",
    "**Note:** In Biology a more common parameter would be the half-life time. The rate $a$ of an event and the half-time $t_{1/2}$ covers the same concept. They are linked by the formula:\n",
    "$$t_{1/2}=\\frac{\\log 2}{a}$$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rateDegradation = 0.1 # min^-1\n",
    "print(\"The mRNA will degrade in %.3f min.\"%numpy.random.exponential(1/rateDegradation)) # caution, it is 1/param"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A very common hypothesis in stochastic modelling of biological event is to represent every event (encounter between molecules, molecule degradations, molecule elongations, etc.) as randomly exponentially distributed variables.There is three main reason to do so:\n",
    "\n",
    "- It is often quite close to the reality (for instance, the degradation of an mRNA).\n",
    "- It is much easier to get analytical formula from such models\n",
    "- <u>It is much simpler to simulate such models</u>\n",
    "\n",
    "Why is it much simpler to simulate by only considering exponentially distributed random variables? Because there is a very usefull trick when using only exponentially distributed random variables that we can use:\n",
    "\n",
    "***\n",
    "**Example**: Let's imagine we have two mRNAs, one is degrading at a rate $0.1\\,min^{-1}$, the other at a rate of $0.3\\,min^{-1}$. We choose to model each of these event with exponentially distributed variables.\n",
    "1. The time WHEN the first of the two degradations happend follows a exponential distribution of parameter $(0.1+0.3)\\,min^{-1}$\n",
    "2. WHICH of the degradation events happen first is given by a random weigthed choice with weights $$\\left(\\frac{0.1}{0.1+0.3},\\frac{0.3}{0.1+0.3}\\right).$$ (that means there is a probability of $\\frac{0.1}{0.1+0.3}$ that the first mRNA degrades at first and $\\frac{0.3}{0.1+0.3}$ that the second mRNA degrades at first.\n",
    "\n",
    "\n",
    "***    \n",
    "\n",
    "This idea is at the core of the Gillespie Algorithm.\n",
    "\n",
    "Let's check the first part (the \"WHEN part\") of the property : \n",
    "\n",
    "***\n",
    "**Exercise:** \n",
    "- create an array `Xarray` of 1000 elements following an exp. distribution of param 0.1\n",
    "- create an array `Yarray` of 1000 elements following an exp. distribution of param 0.3\n",
    "- create an array `minXYarray` of 1000 elements such as `minXYarray[i]` is the minimum between `Xarray[i]` and `Yarray[i]`\n",
    "- create an array `Zarray` of 1000 elements following an exp. distribution of param 0.3+0.1\n",
    "- compare the histograms of `minXYarray` and of `Zarray`\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "################################ TO COMPLETE\n",
    "Xarray = ?????????????????\n",
    "Yarray = ?????????????????\n",
    "minXYarray = ??????????????\n",
    "Zarray = ?????????????????\n",
    "\n",
    "\n",
    "\n",
    "plt.hist(????????????????,normed=True,bins=30,alpha=.5,label='min(X,Y)')\n",
    "plt.hist(????????????????,normed=True,bins=30,alpha=.5,label='Z')\n",
    "plt.legend()\n",
    "##################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's check the second part (the \"WHICH part\") of the Property:\n",
    "\n",
    "***Exercice:***\n",
    "From the previously sampled arrays `Xarray` and `Yarray`, determine the proportion of times an element of `Xarray` happen before `Yarray`, compare this proportion to $\\frac{0.1}{0.1+0.3}$\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "################################ TO COMPLETE\n",
    "\n",
    "?????????????????\n",
    "\n",
    "###########################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The previous property can be generalized to an abritrary number of exponentially distributed random variables:\n",
    "\n",
    "\n",
    "***\n",
    "**Property**: If you have $N$ events modelled by exponentially random variables whose respective rates are $a_1,a_2,\\ldots,a_N$, then we have:\n",
    "1. The time WHEN the first event happen follows a exponential distribution of parameter $a_1,a_2,\\ldots,a_N$.\n",
    "2. WHICH of the events happen first is given by a random weigthed choice with weights $$\\left( \\frac{a_1}{a_1+a_2+\\ldots+a_N},\\frac{a_2}{a_1+a_2+\\ldots+a_N},\\ldots,\\frac{a_N}{a_1+a_2+\\ldots+a_N}\\right).$$ (that means that the probiblity that $i$-th event happen first is $\\frac{a_i}{a_1+a_2+\\ldots+a_N}$)\n",
    "***\n",
    "\n",
    "Why this property will be very usfull for us? Let's imagine we have many possible events (let's say 1000), and times of occurence os each event is modelled as a exp. random variables. How do we determine  which event happen first and when ? We have two equivalent ways:\n",
    "\n",
    "- Method1 (naÃ¯ve): we sample every 1000 exponentially random variables, we then determine which is the minimum\n",
    "- Method2 (using the previous property): we sample 1 exponentially random variable (whose rate is the sum of all the rates) and we randomly picked which event depending on the weight of each event. To use this method, we need the function `randomWeightedChoice` previously defined. \n",
    "\n",
    "Note: In Method2 we will take advantage to the fact that `randomWeightedChoice` does not need normalized weights and directly return the sum of all the weights as a second output.\n",
    "\n",
    "The Method2 is much more efficient as it is shown bellow:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Let's compare the perforance of the two methods:\n",
    "\n",
    "ratesArray = np.random.random(size=1000) # this array will represent the rates of the 1000 random variables\n",
    "\n",
    "def method1():\n",
    "    events = numpy.zeros(1000)\n",
    "    for i,r in enumerate(ratesArray): # for each rate\n",
    "        events[i] = numpy.random.exponential(r) # let's sample a exp random variable according to the rate\n",
    "    \n",
    "    return events.argmin(), events.min()\n",
    "\n",
    "def method2():    \n",
    "    \n",
    "    # We sample WHICH event will happen\n",
    "    event,summ = randomWeightedChoice(ratesArray) # Let's use the function randomWeightedChoice we previously defined\n",
    "    # event is the index that has been chosen\n",
    "    # summ represents the sum of all the weigths (we did not need to normalize `ratesArray`, the function did it for us)\n",
    "    \n",
    "    # We sample WHEN this event will happen\n",
    "    time_step = numpy.random.exponential(1./summ) \n",
    "    return event, time_step\n",
    "\n",
    "# The strange following statements will test the two methods and compute how much time each of them takes.\n",
    "%timeit method1()\n",
    "%timeit method2()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# My first simulation: production and degradation of mRNAs\n",
    "\n",
    "The Gillespeie Algorithm take advantage of the previously explained property to simulate stochastic models very efficiently and with any approximation. It follows the following steps:\n",
    "\n",
    "***\n",
    "**Algorithm**: Gillepsie Algorithm (Gillespie, 1977)\n",
    "\n",
    "    initiation of the state of the simulation\n",
    "    set time =0\n",
    "    while time < T_lim :\n",
    "        determine the rate of all possible events\n",
    "        sample the next event and the time step for the next event to occur\n",
    "        update time occording to the time step\n",
    "        update the state according to the event sampled\n",
    "***\n",
    "        \n",
    "Let's take an example : the production and degradation of mRNAs inside the cell. We can represent this using the following model:\n",
    "\n",
    "Biology | Model\n",
    "-|-\n",
    "<img src=\"./figures/trans.svg\" alt=\"Transcription\" style=\"width: 200px;\"/> |  <img src=\"./figures/TwoStageModel1.png\" alt=\"TwoStageModel1.png\" style=\"width: 200px;\"/><br/>from Dessalles et al. (2017b)\n",
    "\n",
    "The quantity $M$ will be the number of mRNA; it is the state of the simulation. There will be two possible events:\n",
    "- the creation of mRNA. An event of mRNA creation will occur at exp. distributed times of rate $\\lambda_1$. When that even occurs, we increase the number of mRNA $M$ by 1.\n",
    "- the degradation of an mRNA. Each mRNA will degrade after an exp. distributed times of rate $\\sigma_1$. With the previous property, the FIRST time at which ONE of all the mRNAs is degraded is given by a exp. distributed times of rate $\\sigma_1M$. When that even occurs, we decrease the number of mRNA $M$ by 1.\n",
    "\n",
    "***\n",
    "**Exercice**: Let's implement this model using the Gillespie algorithm. The function `simulationTranscription1` takes different paremeters (see the comments), run the Gillespie algorithm and return the number of mRNAs at the end of the simulation. \n",
    "***\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def simulationOneStageModel1(T_lim=1000, lambda1 = 2, sigma1 = 1, M_init = 0):\n",
    "    \"\"\"\n",
    "    Parameters:\n",
    "    - T_lim: the time in minutes of how long is lasting the simulation\n",
    "    - lambda1: the rate of mRNA creation\n",
    "    - sigma1: the rate of mRNA degradation\n",
    "    - M_init: the number of mRNA at the begining of the simulation\n",
    "    \n",
    "    Returns:\n",
    "    - state: the state of the simulation after a simulation of T_lim\n",
    "    \"\"\"\n",
    "    state =  M_init\n",
    "    time = 0\n",
    "    \n",
    "    ################################ TO COMPLETE\n",
    "    while .....................\n",
    "    ?????????????\n",
    "    ########################################\n",
    "    \n",
    "    return state\n",
    "\n",
    "simulationOneStageModel1()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The previous function returns the number of mRNAs after $1000\\,min$ (or whatever value you've choosen for T_lim). It is not that useful. Maybe would be interested seeing how mRNAs are evolving during the cell cycle.\n",
    "\n",
    "For that, we need to regularly save the state of the simulation into a varible that we return at the end of the simulation. We propose to return two quantities:\n",
    "- `saved_time` that will represent each time points at which the simulation has been saved (it depends on the time step `DeltaT` chosen as a parameter).\n",
    "- `saved_state` that will represent what was the value of $M$ at each time point of `saved_time`\n",
    "\n",
    "The algorithm will become:\n",
    "\n",
    "\n",
    "***\n",
    "**Algorithm**: Gillepsie Algorithm (Gillespie, 1977) (with saving)\n",
    "\n",
    "    initialisation of the state of the simulation\n",
    "    ==> initialisation of saving arrays\n",
    "    set time =0\n",
    "    while time < T_lim :\n",
    "        determine the rate of all possible events\n",
    "        sample the next event and the time step for the next event to occur\n",
    "        update time occording to the time step\n",
    "         ==> save the state in the saving arrays if needed\n",
    "        update the state according to the event sampled\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def simulationOneStageModel2(T_lim=1000, lambda1 = 2, sigma1 = 1, M_init = 0, DeltaT = 1. ):\n",
    "    \"\"\"\n",
    "    Parameters:\n",
    "    - T_lim: the time in minute of how long is lasting the simulation\n",
    "    - lambda1: the rate (min^-1) of mRNA creation\n",
    "    - sigma1: the rate (min^-1) of mRNA degradation\n",
    "    - M_init: the number of mRNA at the begining of the simulation\n",
    "    - DeltaT: the time step (in min) at which the simulation is saved\n",
    "    \n",
    "    Returns:\n",
    "    - saved_time: an array that represent each time point for which the state of the simulation has been saved\n",
    "    - saved_state: an array that represent the state of the simulation at the corresponding point in saved_time\n",
    "    \"\"\"\n",
    "    state =  M_init\n",
    "    time = 0\n",
    "    \n",
    "    saved_time = np.arange(0,T_lim,DeltaT)\n",
    "    saved_state = np.zeros(saved_time.shape)\n",
    "    saved_state[0] = state\n",
    "    index_saved_time = 1 # The index we will use to go through saved_time\n",
    "    \n",
    "    \n",
    "    ################################ TO COMPLETE (you can begin by a copy paste of your previous function)\n",
    "    while .....................\n",
    "        ?????????????\n",
    "        ??????????????????\n",
    "         while index_saved_time<len(saved_time) and saved_time[index_saved_time]<= time:\n",
    "            # if we need to save the state of the simulation in saved_state\n",
    "            ???????????????\n",
    "        ???????????????????\n",
    "        ???????????????\n",
    "    \n",
    "    ########################################\n",
    "    \n",
    "    return saved_time,saved_state\n",
    "\n",
    "saved_time,saved_state = simulationOneStageModel2()\n",
    "\n",
    "plt.plot(saved_time,saved_state,label='mRNA number')\n",
    "legend()\n",
    "xlabel('Time [min]')\n",
    "ylabel('Number')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Nicer, isn't ? Take time to play a little bit with the parameters to see what it changes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# A more complicated model: the two stage model\n",
    "\n",
    "Let's then consider model a simple very simple model of protein production: the classical two-stage model (Paulsson, 2005).\n",
    "\n",
    "Biology | Model\n",
    "-|-\n",
    "<img src=\"./figures/transtrad1.svg\" alt=\"Transcription\" style=\"width: 400px;\"/>| <img src=\"./figures/TwoStageModel.png\" alt=\"TwoStageModel.png\" style=\"width: 400px;\"/><br/>from Dessalles et al. (2017b)\n",
    "\n",
    "Now, the state of the simulation is represented by two quantities:\n",
    "- the number of mRNAs $M$\n",
    "- the number of proteins $P$\n",
    "\n",
    "In the simulation, I have chosen to represent the state by a dictionnary (of course, other solutions are possible).\n",
    "\n",
    "\n",
    "And there is also additional events :\n",
    "- the creation of mRNA at rate $\\lambda_1$ (idem as for the previous model)\n",
    "- the degradation of an mRNA at rate $\\sigma_1M$ (idem as for the previous model)\n",
    "- the creation of protein. Each mRNA will create a protein at an exp. distributed time of parameter $\\lambda_2$. Thanks to the property of exponentially distribeted random variables, the FIRST time at which ONE protein is created is given by an exp. distributed time of rate $\\lambda_2P$. When that even occurs, we increase the number of mRNA $P$ by 1.\n",
    "- proteolysis. Each protein will degrade after an exp. distributed times of rate $\\sigma_2$.So, the FIRST time at which ONE of all the proteins is degraded is given by a exp. distributed times of rate $\\sigma_2P$. When that even occurs, we decrease the number of mRNA $P$ by 1.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def simulationTwoStageModel(T_lim=1000, \n",
    "                        lambda1 = 2, sigma1 = 1, M_init = 0, \n",
    "                        lambda2 = 10, sigma2 = 0.011, P_init = 0, \n",
    "                        DeltaT = 1. ):\n",
    "    \"\"\"\n",
    "    Parameters:\n",
    "    - T_lim: the time in minute of how long will last the simulation\n",
    "    - lambda1: the rate (min^-1) of mRNA creation\n",
    "    - sigma1: the rate (min^-1) of mRNA degradation\n",
    "    - M_init: the number of mRNA at the begining of the simulation\n",
    "    - lambda2: the rate (min^-1) of protein creation per mRNA\n",
    "    - sigma2: the rate (min^-1) of protein proteolysis\n",
    "    - P_init: the number of mRNA at the begining of the simulation\n",
    "    - DeltaT: the time step (in min) at which the simulation is saved\n",
    "    \n",
    "    Returns:\n",
    "    - saved_time: an array that represent each time point for which the state of the simulation has been saved\n",
    "    - saved_state: an array that represent the state of the simulation at the corresponding point in saved_time\n",
    "    \"\"\"\n",
    "    state =  {'M':M_init,'P':P_init} # the state will be represented here as a dictionnary\n",
    "    time = 0\n",
    "    \n",
    "    saved_time = np.arange(0,T_lim,DeltaT)\n",
    "    saved_state = {'M':np.zeros(saved_time.shape),'P':np.zeros(saved_time.shape)}\n",
    "    saved_state['M'][0] = state['M'] \n",
    "    saved_state['P'][0] = state['P'] \n",
    "    index_saved_time = 1 # The index we will use to go through saved_time\n",
    "    \n",
    "    ################################ TO COMPLETE (you can begin by a copy paste of your previous function)\n",
    "    while .....................\n",
    "    ?????????????\n",
    "    ########################################\n",
    "    \n",
    "    return saved_time,saved_state\n",
    "\n",
    "saved_time,saved_state = simulationTwoStageModel()\n",
    "\n",
    "figure()\n",
    "plt.plot(saved_time,saved_state['M'],label='mRNA number')\n",
    "legend()\n",
    "xlabel('Time [min]')\n",
    "ylabel('Number of mRNA')\n",
    "figure()\n",
    "plt.plot(saved_time,saved_state['P'],color='r',label='protein number')\n",
    "legend()\n",
    "xlabel('Time [min]')\n",
    "ylabel('Number of Proteins')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can now begin the project with the information given at the begining of the Notebook.\n",
    "\n",
    "# References\n",
    "\n",
    "- Thisted, Thomas, and Kenn Gerdes. \"Mechanism of post-segregational killing by the hok/sok system of plasmid R1: Sok antisense RNA regulates hok gene expression indirectly through the overlapping mok gene.\" Journal of molecular biology 223.1 (1992): 41-54.\n",
    "- Franch, Thomas, Alexander P. Gultyaev, and Kenn Gerdes. \"Programmed cell death by hok/sok of plasmid R1: Processing at the hok mRNA 3â€²-end triggers structural rearrangements that allow translation and antisense RNA binding11Edited by DE Draper.\" Journal of molecular biology 273, no. 1 (1997): 38-51.\n",
    "- Gillespie, D.T. (1977). Exact stochastic simulation of coupled chemical reactions. J. Phys. Chem. 81, 2340â€“2361.\n",
    "- Paulsson, J. (2005). Models of stochastic gene expression. Physics of Life Reviews 2, 157â€“175.\n",
    "- Dessalles, R., Fromion, V., and Robert, P. (2017a). A stochastic analysis of autoregulation of gene expression. J. Math. Biol. 1â€“31.\n",
    "- Dessalles, R., Fromion, V., and Robert, P. (2017b). Models of protein production with cell cycle. ArXiv:1711.06378 [q-Bio].\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
